### ISSUE TITLE
[Deprecation] Properly deprecate CentOS7 as CI build image/Supported OS and switch to Almalinux8

### ISSUE THREAD

**Issue by peterzhuamazon** [opensearch-project/opensearch-build#4379]
###### 20240430 We will switch back to CentOS7 for another release in 2.14.0

**[Deprecation] Properly deprecate CentOS7 as CI build image/Supported OS and switch to RockyLinux8**

Back in #3743, we already plan to migrate all CentOS7 images to RockyLinux8 as part of the NodeJS18(#1563) / Python3.9(#3351) Upgrade.

However, it was pushed back by the k-NN Team as moving from CentOS7 to RockyLinux8 would change the minimal glibc version from 2.17 to 2.28 during lib build, causing people with servers containing glibc lower than 2.28 failed to run k-NN, as the plugin will crash upon api call.
* https://github.com/opensearch-project/opensearch-build/pull/4002

CentOS7 will end its support by 2024/06/30 (https://endoflife.date/centos), meaning we need to migrate to RockyLinux8 by then, especially for the only build image for OpenSearch. We also need to properly announce it on the website, as our compatibility chart is clearly mentioning the support for it here: https://github.com/opensearch-project/documentation-website/blob/054dc56d5dcef66576cd4b15bef0ae5efa4f8f77/_install-and-configure/install-opensearch/index.md?plain=1#L26

We need to have a plan and decide when do we announce:
1. Discontinue CentOS7 support on Release
2. Discontinue CentOS7 support on CI Build
  2.1. Which in turn, k-NN drop CentOS7 support on its lib as well

Thanks.

------

* More related: https://github.com/opensearch-project/opensearch-build/issues/4687

------

PRs:
* https://github.com/opensearch-project/opensearch-build/pull/4464
* https://github.com/opensearch-project/opensearch-build/pull/4645
* https://github.com/opensearch-project/opensearch-build/pull/4647
* https://github.com/opensearch-project/k-NN/pull/1634
* https://github.com/opensearch-project/opensearch-build/pull/4663
* https://github.com/opensearch-project/k-NN/pull/1664
* https://github.com/opensearch-project/opensearch-build/pull/4667
* https://github.com/opensearch-project/k-NN/pull/1674
* https://github.com/opensearch-project/opensearch-build/pull/4714
* https://github.com/opensearch-project/documentation-website/pull/7287
* https://github.com/opensearch-project/opensearch-build/pull/4744
* https://github.com/opensearch-project/opensearch-build/pull/4760
* https://github.com/opensearch-project/opensearch-build/pull/4809
* #4820 
* https://github.com/opensearch-project/opensearch-build/pull/4903

**Comment by bbarani** [comment]
@vamshin @jmazanec15 @VijayanB We need to put out deprecation notice before we stop support for CentOS7. CC: @krisfreedain

**Comment by krisfreedain** [comment]
@vamshin @jmazanec15 @VijayanB - Agree with @bbarani - I'd recommend this being included in the Release Notes, and likely the launch blog @jhmcintyre

**Comment by peterzhuamazon** [comment]
We plan to switch to Almalinux from Rockylinux soon.
https://github.com/opensearch-project/opensearch-build/pull/4525

**Comment by jmazanec15** [comment]
@peterzhuamazon What glibc does almalinux have?

**Comment by bbarani** [comment]
Tagging @elfisher @krisfreedain as we are nearing CentOS 7 EOL date (June 30, 2024).

**Comment by peterzhuamazon** [comment]
After offline discussion with k-NN team, we will now switch the 2.14.0 manifest to use Almalinux8 for building OpenSearch.

cc: @vamshin @jmazanec15 @bbarani 

Thanks.

**Comment by peterzhuamazon** [comment]
Seems like there are some issues with almalinux8 running gcc 9.2.1 for ARM64 build.
@jmazanec15 has reported to me and he will add more details.
I will try to raise from gcc 9.2.1 to 9.3.1 which is the version we used back in CentOS7 soon.

**Comment by jmazanec15** [comment]
@peterzhuamazon right - issue here: https://github.com/opensearch-project/k-NN/issues/1591#issuecomment-2066809899.

**Comment by peterzhuamazon** [comment]
Based on https://gcc.gnu.org/bugzilla/show_bug.cgi?id=94052, it seems the bug has been fixed in gcc 10 then backport to gcc 9 and gcc 8 on 2020/03/24.

According to https://gcc.gnu.org/releases.html, 9.2 released on 2019/08/12 thus not containing the fix until 9.3 in 2020/03.

A few routes:
* Get 9.3.1 source code and compile 9.3.1 just for rockylinux8.
* Go to gcc10.
* Go to gcc12 as Windows is already using that version for a long time.

Windows:
```
$ which gcc
/c/Users/ContainerAdministrator/scoop/apps/mingw/12.2.0-rt_v10-rev1/bin/g++

$ g++ --version
g++.exe (x86_64-posix-seh-rev1, Built by MinGW-W64 project) 12.2.0
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

After offline discussion with @jmazanec15 we have no concern going to gcc12.

Will PR soon.

Thanks.

**Comment by peterzhuamazon** [comment]
More on CentOS7 replacement if going to Rockylinux8 is too much:
https://www.redhat.com/en/topics/linux/centos-alternatives

**Comment by peterzhuamazon** [comment]
One more issue:
https://github.com/OpenMathLib/OpenBLAS/issues/3708
https://sourceware.org/bugzilla/show_bug.cgi?id=29435

Seems like issue due to gcc 11.2.1 is bundled with dep of binutils 2.38.
We will switch back to 10.3.1 gcc and default binutils 2.3.0, @jmazanec15 will verify.

Thanks.

**Comment by peterzhuamazon** [comment]
Apparently gcc10 still have bug on arm64, go back to gcc11 for now.

```
[  2%] Linking CXX shared library release/libopensearchknn_common.so
[  2%] Built target opensearchknn_common
[  2%] Building CXX object external/nmslib/similarity_search/src/CMakeFiles/NonMetricSpaceLib.dir/distcomp_bregman.cc.o
In file included from /home/ci-runner/k-NN/jni/external/nmslib/similarity_search/src/distcomp_bregman.cc:17:
/home/ci-runner/k-NN/jni/external/nmslib/similarity_search/src/distcomp_bregman.cc: In function 'T similarity::ItakuraSaitoPrecompSIMD(const T*, const T*, size_t) [with T = float; size_t = long unsigned int]':
/home/ci-runner/k-NN/jni/external/nmslib/similarity_search/include/utils.h:53:53: note: '#pragma message: /home/ci-runner/k-NN/jni/external/nmslib/similarity_search/src/distcomp_bregman.cc(83) : WARNING: ItakuraSaitoPrecompSIMD<float>: SSE2 is not available, defaulting to pure C++ implementation!'
   53 | #   define WARN(exp) (FILE_LINE_LINK "WARNING: " exp)
      |                                                     ^
/home/ci-runner/k-NN/jni/external/nmslib/similarity_search/src/distcomp_bregman.cc:83:17: note: in expansion of macro 'WARN'
   83 | #pragma message WARN("ItakuraSaitoPrecompSIMD<float>: SSE2 is not available, defaulting to pure C++ implementation!")
      |                 ^~~~
/home/ci-runner/k-NN/jni/external/nmslib/similarity_search/src/distcomp_bregman.cc: In function 'T similarity::KLPrecompSIMD(const T*, const T*, size_t) [with T = float; size_t = long unsigned int]':
/home/ci-runner/k-NN/jni/external/nmslib/similarity_search/include/utils.h:53:53: note: '#pragma message: /home/ci-runner/k-NN/jni/external/nmslib/similarity_search/src/distcomp_bregman.cc(217) : WARNING: KLPrecompSIMD<float>: SSE2 is not available, defaulting to pure C++ implementation!'
   53 | #   define WARN(exp) (FILE_LINE_LINK "WARNING: " exp)
      |                                                     ^
/home/ci-runner/k-NN/jni/external/nmslib/similarity_search/src/distcomp_bregman.cc:217:17: note: in expansion of macro 'WARN'
  217 | #pragma message WARN("KLPrecompSIMD<float>: SSE2 is not available, defaulting to pure C++ implementation!")
      |                 ^~~~
/home/ci-runner/k-NN/jni/external/nmslib/similarity_search/src/distcomp_bregman.cc: In function 'T similarity::KLGeneralPrecompSIMD(const T*, const T*, size_t) [with T = float; size_t = long unsigned int]':
/home/ci-runner/k-NN/jni/external/nmslib/similarity_search/include/utils.h:53:53: note: '#pragma message: /home/ci-runner/k-NN/jni/external/nmslib/similarity_search/src/distcomp_bregman.cc(329) : WARNING: KLGeneralPrecompSIMD<float>: SSE2 is not available, defaulting to pure C++ implementation!'
   53 | #   define WARN(exp) (FILE_LINE_LINK "WARNING: " exp)
      |                                                     ^
/home/ci-runner/k-NN/jni/external/nmslib/similarity_search/src/distcomp_bregman.cc:329:17: note: in expansion of macro 'WARN'
  329 | #pragma message WARN("KLGeneralPrecompSIMD<float>: SSE2 is not available, defaulting to pure C++ implementation!")
      |                 ^~~~
Assembler messages:
Error: unknown architectural extension `ssbs'
Error: unrecognized option -march=armv8.2-a+crypto+fp16+rcpc+dotprod+ssbs
make[2]: *** [external/nmslib/similarity_search/src/CMakeFiles/NonMetricSpaceLib.dir/build.make:76: external/nmslib/similarity_search/src/CMakeFiles/NonMetricSpaceLib.dir/distcomp_bregman.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:457: external/nmslib/similarity_search/src/CMakeFiles/NonMetricSpaceLib.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

**Comment by peterzhuamazon** [comment]
We will switch back to CentOS7 for another release in 2.14.0.

**Comment by peterzhuamazon** [comment]
More discussion with @jmazanec15 on if we want to separate the lib build process out of the main distribution pipeline:
We'd have to have a separate image for ppc64le then. Then we need:
* AL2 for linux arm64 and x64
* Alma for ppc64le
* Windows
Then, build would need to: compile the libs, run jni specific test suite, publish to s3 bucket

If we go this route then ppc64le would have to start with alma8 as baseline.

**Comment by peterzhuamazon** [comment]
More discussions on the next steps of k-NN lib building process:
* https://github.com/opensearch-project/opensearch-build/issues/4687

**Comment by peterzhuamazon** [comment]
More PRs:
* #4714

**Comment by peterzhuamazon** [comment]
More issues related to the arm64 version of al2:
```

#43 [linux/arm64 16/21] RUN yum repolist && yum install openblas-static lapack gcc-gfortran -y
#43 2.491 Loaded plugins: ovl, priorities
#43 24.30 227 packages excluded due to repository priority protections
#43 24.49 repo id                   repo name                                    status
#43 24.49 amzn2-core/2/aarch64      Amazon Linux 2 core repository                   25612
#43 24.49 amzn2extra-epel/2/aarch64 Amazon Extras repo for epel                          1
#43 24.49 epel/aarch64              Extra Packages for Enterprise Linux 7 - aarc 12761+227
#43 24.49 repolist: 38374
#43 25.69 Loaded plugins: ovl, priorities
#43 32.99 227 packages excluded due to repository priority protections
#43 41.07 Package gcc-gfortran-7.3.1-17.amzn2.aarch64 already installed and latest version
#43 41.07 Resolving Dependencies
#43 41.07 --> Running transaction check
#43 41.07 ---> Package lapack.aarch64 0:3.4.2-8.amzn2.0.2 will be installed
#43 41.10 --> Processing Dependency: libblas.so.3()(64bit) for package: lapack-3.4.2-8.amzn2.0.2.aarch64
#43 41.71 ---> Package openblas-static.aarch64 0:0.3.3-2.el7 will be installed
#43 41.71 --> Processing Dependency: openblas-devel(aarch-64) = 0.3.3-2.el7 for package: openblas-static-0.3.3-2.el7.aarch64
#43 41.72 --> Running transaction check
#43 41.72 ---> Package blas.aarch64 0:3.4.2-8.amzn2.0.2 will be installed
#43 41.72 ---> Package openblas-devel.aarch64 0:0.3.3-2.el7 will be installed
#43 41.73 --> Processing Dependency: openblas(aarch-64) = 0.3.3-2.el7 for package: openblas-devel-0.3.3-2.el7.aarch64
#43 41.73 --> Processing Dependency: openblas-openmp(aarch-64) = 0.3.3-2.el7 for package: openblas-devel-0.3.3-2.el7.aarch64
#43 41.73 --> Processing Dependency: openblas-openmp64(aarch-64) = 0.3.3-2.el7 for package: openblas-devel-0.3.3-2.el7.aarch64
#43 41.74 --> Processing Dependency: openblas-openmp64_(aarch-64) = 0.3.3-2.el7 for package: openblas-devel-0.3.3-2.el7.aarch64
#43 41.75 --> Processing Dependency: openblas-serial(aarch-64) = 0.3.3-2.el7 for package: openblas-devel-0.3.3-2.el7.aarch64
#43 41.75 --> Processing Dependency: openblas-serial64(aarch-64) = 0.3.3-2.el7 for package: openblas-devel-0.3.3-2.el7.aarch64
#43 41.75 --> Processing Dependency: openblas-serial64_(aarch-64) = 0.3.3-2.el7 for package: openblas-devel-0.3.3-2.el7.aarch64
#43 41.76 --> Processing Dependency: openblas-threads(aarch-64) = 0.3.3-2.el7 for package: openblas-devel-0.3.3-2.el7.aarch64
#43 41.76 --> Processing Dependency: openblas-threads64(aarch-64) = 0.3.3-2.el7 for package: openblas-devel-0.3.3-2.el7.aarch64
#43 41.77 --> Processing Dependency: openblas-threads64_(aarch-64) = 0.3.3-2.el7 for package: openblas-devel-0.3.3-2.el7.aarch64
#43 41.77 --> Processing Dependency: libopenblaso.so.0()(64bit) for package: openblas-devel-0.3.3-2.el7.aarch64
#43 41.77 --> Processing Dependency: libopenblaso64.so.0()(64bit) for package: openblas-devel-0.3.3-2.el7.aarch64
#43 41.78 --> Processing Dependency: libopenblaso64_.so.0()(64bit) for package: openblas-devel-0.3.3-2.el7.aarch64
#43 41.78 --> Running transaction check
#43 41.78 ---> Package openblas.aarch64 0:0.3.3-2.el7 will be installed
#43 41.78 ---> Package openblas-openmp.aarch64 0:0.3.3-2.el7 will be installed
#43 41.79 --> Processing Dependency: libgfortran.so.3(GFORTRAN_1.0)(64bit) for package: openblas-openmp-0.3.3-2.el7.aarch64
#43 41.80 --> Processing Dependency: libgfortran.so.3()(64bit) for package: openblas-openmp-0.3.3-2.el7.aarch64
#43 41.80 ---> Package openblas-openmp64.aarch64 0:0.3.3-2.el7 will be installed
#43 41.80 --> Processing Dependency: libgfortran.so.3(GFORTRAN_1.0)(64bit) for package: openblas-openmp64-0.3.3-2.el7.aarch64
#43 41.80 --> Processing Dependency: libgfortran.so.3()(64bit) for package: openblas-openmp64-0.3.3-2.el7.aarch64
#43 41.81 ---> Package openblas-openmp64_.aarch64 0:0.3.3-2.el7 will be installed
#43 41.81 --> Processing Dependency: libgfortran.so.3(GFORTRAN_1.0)(64bit) for package: openblas-openmp64_-0.3.3-2.el7.aarch64
#43 41.81 --> Processing Dependency: libgfortran.so.3()(64bit) for package: openblas-openmp64_-0.3.3-2.el7.aarch64
#43 41.81 ---> Package openblas-serial.aarch64 0:0.3.3-2.el7 will be installed
#43 41.82 --> Processing Dependency: libgfortran.so.3(GFORTRAN_1.0)(64bit) for package: openblas-serial-0.3.3-2.el7.aarch64
#43 41.82 --> Processing Dependency: libgfortran.so.3()(64bit) for package: openblas-serial-0.3.3-2.el7.aarch64
#43 41.82 ---> Package openblas-serial64.aarch64 0:0.3.3-2.el7 will be installed
#43 41.82 --> Processing Dependency: libgfortran.so.3(GFORTRAN_1.0)(64bit) for package: openblas-serial64-0.3.3-2.el7.aarch64
#43 41.82 --> Processing Dependency: libgfortran.so.3()(64bit) for package: openblas-serial64-0.3.3-2.el7.aarch64
#43 41.83 ---> Package openblas-serial64_.aarch64 0:0.3.3-2.el7 will be installed
#43 41.83 --> Processing Dependency: libgfortran.so.3(GFORTRAN_1.0)(64bit) for package: openblas-serial64_-0.3.3-2.el7.aarch64
#43 41.83 --> Processing Dependency: libgfortran.so.3()(64bit) for package: openblas-serial64_-0.3.3-2.el7.aarch64
#43 41.83 ---> Package openblas-threads.aarch64 0:0.3.3-2.el7 will be installed
#43 41.83 --> Processing Dependency: libgfortran.so.3(GFORTRAN_1.0)(64bit) for package: openblas-threads-0.3.3-2.el7.aarch64
#43 41.84 --> Processing Dependency: libgfortran.so.3()(64bit) for package: openblas-threads-0.3.3-2.el7.aarch64
#43 41.84 ---> Package openblas-threads64.aarch64 0:0.3.3-2.el7 will be installed
#43 41.84 --> Processing Dependency: libgfortran.so.3(GFORTRAN_1.0)(64bit) for package: openblas-threads64-0.3.3-2.el7.aarch64
#43 41.84 --> Processing Dependency: libgfortran.so.3()(64bit) for package: openblas-threads64-0.3.3-2.el7.aarch64
#43 41.85 ---> Package openblas-threads64_.aarch64 0:0.3.3-2.el7 will be installed
#43 41.85 --> Processing Dependency: libgfortran.so.3(GFORTRAN_1.0)(64bit) for package: openblas-threads64_-0.3.3-2.el7.aarch64
#43 41.85 --> Processing Dependency: libgfortran.so.3()(64bit) for package: openblas-threads64_-0.3.3-2.el7.aarch64
#43 43.24 --> Finished Dependency Resolution
#43 43.24 Error: Package: openblas-openmp-0.3.3-2.el7.aarch64 (epel)
#43 43.24            Requires: libgfortran.so.3(GFORTRAN_1.0)(64bit)
#43 43.24 Error: Package: openblas-openmp-0.3.3-2.el7.aarch64 (epel)
#43 43.24            Requires: libgfortran.so.3()(64bit)
#43 43.24 Error: Package: openblas-serial-0.3.3-2.el7.aarch64 (epel)
#43 43.24            Requires: libgfortran.so.3(GFORTRAN_1.0)(64bit)
#43 43.24 Error: Package: openblas-serial64_-0.3.3-2.el7.aarch64 (epel)
#43 43.24            Requires: libgfortran.so.3()(64bit)
#43 43.24 Error: Package: openblas-serial64_-0.3.3-2.el7.aarch64 (epel)
#43 43.24            Requires: libgfortran.so.3(GFORTRAN_1.0)(64bit)
#43 43.25 Error: Package: openblas-openmp64_-0.3.3-2.el7.aarch64 (epel)
#43 43.25            Requires: libgfortran.so.3()(64bit)
#43 43.25 Error: Package: openblas-threads64-0.3.3-2.el7.aarch64 (epel)
#43 43.25            Requires: libgfortran.so.3()(64bit)
#43 43.25 Error: Package: openblas-serial64-0.3.3-2.el7.aarch64 (epel)
#43 43.25            Requires: libgfortran.so.3(GFORTRAN_1.0)(64bit)
#43 43.25 Error: Package: openblas-threads64_-0.3.3-2.el7.aarch64 (epel)
#43 43.25            Requires: libgfortran.so.3(GFORTRAN_1.0)(64bit)
#43 43.25 Error: Package: openblas-openmp64-0.3.3-2.el7.aarch64 (epel)
#43 43.25            Requires: libgfortran.so.3(GFORTRAN_1.0)(64bit)
#43 43.25 Error: Package: openblas-openmp64-0.3.3-2.el7.aarch64 (epel)
#43 43.25            Requires: libgfortran.so.3()(64bit)
#43 43.25 Error: Package: openblas-threads-0.3.3-2.el7.aarch64 (epel)
#43 43.25            Requires: libgfortran.so.3()(64bit)
#43 43.25 Error: Package: openblas-serial64-0.3.3-2.el7.aarch64 (epel)
#43 43.25            Requires: libgfortran.so.3()(64bit)
#43 43.25 Error: Package: openblas-threads-0.3.3-2.el7.aarch64 (epel)
#43 43.25            Requires: libgfortran.so.3(GFORTRAN_1.0)(64bit)
#43 43.25 Error: Package: openblas-threads64-0.3.3-2.el7.aarch64 (epel)
#43 43.25            Requires: libgfortran.so.3(GFORTRAN_1.0)(64bit)
#43 43.25 Error: Package: openblas-openmp64_-0.3.3-2.el7.aarch64 (epel)
#43 43.25            Requires: libgfortran.so.3(GFORTRAN_1.0)(64bit)
#43 43.25 Error: Package: openblas-threads64_-0.3.3-2.el7.aarch64 (epel)
#43 43.25            Requires: libgfortran.so.3()(64bit)
#43 43.25 Error: Package: openblas-serial-0.3.3-2.el7.aarch64 (epel)
#43 43.25            Requires: libgfortran.so.3()(64bit)
#43 43.25  You could try using --skip-broken to work around the problem
#43 44.35  You could try running: rpm -Va --nofiles --nodigest
#43 ERROR: process "/bin/bash -lc yum repolist && yum install openblas-static lapack gcc-gfortran -y" did not complete successfully: exit code: 1
------
 > [linux/arm64 16/21] RUN yum repolist && yum install openblas-static lapack gcc-gfortran -y:
#43 43.25 Error: Package: openblas-threads64-0.3.3-2.el7.aarch64 (epel)
#43 43.25            Requires: libgfortran.so.3(GFORTRAN_1.0)(64bit)
#43 43.25 Error: Package: openblas-openmp64_-0.3.3-2.el7.aarch64 (epel)
#43 43.25            Requires: libgfortran.so.3(GFORTRAN_1.0)(64bit)
#43 43.25 Error: Package: openblas-threads64_-0.3.3-2.el7.aarch64 (epel)
#43 43.25            Requires: libgfortran.so.3()(64bit)
#43 43.25 Error: Package: openblas-serial-0.3.3-2.el7.aarch64 (epel)
#43 43.25            Requires: libgfortran.so.3()(64bit)
#43 43.25  You could try using --skip-broken to work around the problem
#43 44.35  You could try running: rpm -Va --nofiles --nodigest
```

**Comment by peterzhuamazon** [comment]
ARM64 AL2 is missing some certain packages:
```

bash-4.2#  yum install -y compat-gcc-48-libgfortran
Loaded plugins: ovl, priorities
6945 packages excluded due to repository priority protections
No package compat-gcc-48-libgfortran available.
Error: Nothing to do
```

**Comment by peterzhuamazon** [comment]
It is now confirmed based on the al2 pkg release notes:
`- Disable Arm builds. They were never distributed and aren't supported.`

**Comment by peterzhuamazon** [comment]
Seems like force install compat would not be a good idea, but force install openblas manually one by one with cpio might be better, as yum is requesting glibc as a dependency pkg even though it is already there.

Also libquadmath is also looking for the glibc during installation.

**Comment by peterzhuamazon** [comment]
Seems like building from the srpm is a no go at this point, due to it is looking for `/lib/glibc.so.6` instead of `/lib/lib64/glibc.so.6`.

**Comment by peterzhuamazon** [comment]
Having discussed with @junqiu-lei and there are a few approaches here:
1. Trying to get AL2 ARM64 install with all the necessary libs, do not use yum directly but rather manually/automated downloading all the rpm packages and use cpio to extract to corresponding directories.
2. Trying to cross-compile knnlib on AL2 x64 for ARM64, need to check the cmake and make file changes.
3. Trying to use non-RH flavored OS to compile, such as Ubuntu 16/18, but need kNN team support on the equivalent packages of `openblas-static lapack gfortran`.
4. Drop AL2 glibc2.26 support and move on to Almalinux8 with glibc2.28.

Note that AL2 is bundled with gcc7 and gcc10, which both have issues in the past and we are forced to migrate to higher version in the past. We are not clear whether the bundled version on AL2 can compile knnlib correctly.

**Comment by junqiu-lei** [comment]
I tried above option1 which installed all the necessary libs from source, I found the error only happened in [jni_test part](https://github.com/opensearch-project/k-NN/blob/main/jni/CMakeLists.txt#L129-L182) with:
```
[ 83%] Linking CXX executable bin/jni_test
external/faiss/faiss/libfaiss.a(IndexAdditiveQuantizerFastScan.cpp.o): In function `faiss::IndexAdditiveQuantizerFastScan::search(long, float const*, long, float*, long*, faiss::SearchParameters const*) const':
IndexAdditiveQuantizerFastScan.cpp:(.text+0xc88): undefined reference to `void faiss::IndexFastScan::search_dispatch_implem<true>(long, float const*, long, float*, long*, faiss::NormTableScaler const*) const'
release/libopensearchknn_faiss.so: undefined reference to `_gfortran_concat_string'
collect2: error: ld returned 1 exit status
make[2]: *** [bin/jni_test] Error 1
``` 

After I removed the jni_test part and run `./gradlew integTest`, all integration test can be passed:
```
> Task :integTest
May 22, 2024 5:09:20 PM sun.util.locale.provider.LocaleProviderAdapter <clinit>
WARNING: COMPAT locale provider will be removed in a future release

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.4/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 26m 7s
19 actionable tasks: 19 executed
[ci-runner@a26d5cd0a0e9 k-NN]$ arch
aarch64
```

So maybe we can bypass the jni_test part for AL2 arm64 env build.

Ref for the dockerfile diff: https://github.com/opensearch-project/opensearch-build/compare/main...junqiu-lei:opensearch-build:knn-build-1

**Comment by peterzhuamazon** [comment]
Hi @junqiu-lei ,
A few comments based on this commit: https://github.com/opensearch-project/opensearch-build/compare/main...junqiu-lei:opensearch-build:knn-build-1#diff-3ed70988a2fbf3a379c3afd4b908a4f434627c38c5c472fc8be17acf3d5654a2

1. We would consider actually using the 0.3.3 branch for building, as that is the version available on AL2 / CentOS7.
https://github.com/OpenMathLib/OpenBLAS/tree/v0.3.3
2. Keep lapack and gfortran but make both x64 and arm64 using build from source openblas, remove `TARGET` and let openblas automatically decide on the build target, need testing on both x64 and arm64.
```
RUN yum repolist && yum install lapack gcc-gfortran -y && \
        git clone -b v0.3.3 --single-branch https://github.com/xianyi/OpenBLAS.git && \
        cd OpenBLAS && \
        make USE_OPENMP=1 FC=gfortran && \
        make PREFIX=/usr/local install

ENV LD_LIBRARY_PATH="/usr/local/lib:$LD_LIBRARY_PATH"
```
3. If needed, can verify whether jni test compilation can be fixed, but not blocker.

Thanks.

**Comment by peterzhuamazon** [comment]
Adding exist criteria based on Jack recommendations:
1. Create build images based on AL2
2. Test the compiled lib files on both x64 and arm64 of AL2, AL2023, Ubuntu20.04, Alma linux8.
3. Test the compiled lib files as part of the docker distribution
4. Add the lib generation part by tweaking the existing distribution build process so it is standalone stage.
Thanks.

**Comment by peterzhuamazon** [comment]
We are now replacing the gcc7 with gcc10 using alternatives.
Even though gcc still shows as 7 in compilation, we are able to compile avx2 with no issues and confirms it is actually running the gcc10 binary.

Next step:
* @junqiu-lei send a PR with the changes discussed in the call
* @peterzhuamazon  approve / merge the changes and build the new AL2 image
* @junqiu-lei  test the new changes on multiple OSes to confirm their functionalities.

**Comment by junqiu-lei** [comment]
After upgrading openblas version to v0.3.27 from source install and use GCC 10.5, we're able to build artifact and complete remote integ test on both arm64 and x64 AL2 envs. #4735

**Comment by peterzhuamazon** [comment]
New build with AL2 running knnlib build: https://build.ci.opensearch.org/job/distribution-build-opensearch/9905/console

@junqiu-lei will test and see if we can proceed to the next step.

**Comment by junqiu-lei** [comment]
With the build CI artifacts triggered by Peter, all the remote integ tests passed on X64 and Arm64 AL2 envs。

**Comment by peterzhuamazon** [comment]
Related docs PR:
* https://github.com/opensearch-project/documentation-website/pull/7287

**Comment by peterzhuamazon** [comment]
One more:
* https://github.com/opensearch-project/opensearch-build/pull/4744

**Comment by peterzhuamazon** [comment]
Due to additional issues with the lib we are reverting back to CentOS7 for another release:
* https://github.com/opensearch-project/k-NN/issues/1737

**Comment by Pallavi-AWS** [comment]
> Due to additional issues with the lib we are reverting back to CentOS7 for another release:
> 
>     * [[BUG] CI failing on Linux with Java runtime error k-NN#1737](https://github.com/opensearch-project/k-NN/issues/1737)

Thanks @peterzhuamazon. We would need to prioritize the deprecation in next release though - @vamshin @navneet1v

**Comment by junqiu-lei** [comment]
In AL2 docker image we build openblas from source from [here](https://github.com/opensearch-project/opensearch-build/blob/main/docker/ci/dockerfiles/current/build.al2.opensearch.x64.arm64.dockerfile#L105-L108) through build team pipeline with m5.4xlarge machine.

The above additional issue was found when we test against on different arch on x64 architecture. For example, it can be reproduced when run knn test on c5a.4xlarge machine with default docker image build from m5.4xlarge.

After adding `DYNAMIC_ARCH=1` argument to openblas make, we found the issue can be resolve in above test example, we still need to verify the change in other envs(arm64/x64 on AL2, AL2023, Ubuntu 20.04).

**Comment by naveentatikonda** [comment]
Had discussion with a couple of folks from Intel team for adding AVX512 support. There are a couple of requirements that we have identified so far: 

- gcc compiler should be atleast of version 12 
- library needs to be built on 4th Gen Intel® Xeon® Scalable processor (like the r7i )

One blocker is yum only supports gcc versions 7 and 10 on AL2 (as we are switching to AL2 from OpenSearch 2.16), so we might need to manually compile it. @peterzhuamazon please let me know if you think there are any other blockers from build  POV.
cc: @vamshin @navneet1v @jmazanec15 @junqiu-lei

**Comment by peterzhuamazon** [comment]
Restart this process again to have k-NN work in 2.16.0, tagging @prudhvigodithi as he would be the 2.16.0 release manager.

I will proceed the next steps with @junqiu-lei 
1. Junqiu push the changes of using `DYNAMIC_ARCH=1` to AL2 image for x64 only
2. Peter proceed with rebuilding the AL2 image and publish to all registries
3. Peter proceed with the docker image switch again to AL2
4. Peter make changes to the centos7 deprecation documentation PRs and highlight text for 2.16.0

Thanks.

**Comment by peterzhuamazon** [comment]
Try switch again from centos7 to al2 with dynamic_arch changes:
* https://github.com/opensearch-project/opensearch-build/pull/4809

**Comment by peterzhuamazon** [comment]
Thanks @junqiu-lei ,

We are moving this issue to backlog as the AL2 part is completed.
We will re-visit this once the Almalinux8 part is coming next year. Thanks.

**Comment by peterzhuamazon** [comment]
New PR:
* https://github.com/opensearch-project/opensearch-build/pull/4820

**Comment by dbwiddis** [comment]
> CentOS7 will end its support by 2024/06/30 (https://endoflife.date/centos), meaning we need to migrate to RockyLinux8 by then

Hello from 2024/07/06! :) 

GitHub-based actions using this container no longer work.  GitHub switched their runners to Node20 [as of June 3](https://github.blog/changelog/2024-03-07-github-actions-all-actions-will-run-on-node20-instead-of-node16-by-default/) which have eventually rolled out to the runners for plugin CI.

TLDR: It looks like PR #4820 is intended to address this, but the issue also mentions NodeJS 18.  Will that PR support NodeJS 20, which is now a minimal version for any GitHub Actions?

**Comment by peterzhuamazon** [comment]
Note that we are are shipping nodejs18 on OSD. (edited) 

This is particularly that opensearch plugins repos are currently running on AL2 build images for checks purposes.

And github checkout action requires to install node to checkout.


This should have nothing to do with your code compilation or anything, it is just for the sake of run that action to git checkout code.


OSD plugins does not affect by this at all.


Because we are only supporting AL2 for another year, for support k-NN native code compilation on glibc 2.26.


Hope that can explain the confusion :sweat_smile:  Thanks.

**Comment by peterzhuamazon** [comment]
Set 3.0.0 AL2:
* https://github.com/opensearch-project/opensearch-build/pull/4903

**Comment by peterzhuamazon** [comment]
This has been completed and will be followed by:
* #2838
